File: src/lib/tauri/events.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import { isTauri } from "./invoke";

export const TAURI_EVENTS = {
  agentMessage: "agent:message",
  agentStatus: "agent:status",
  agentToolStart: "agent:tool-start",
  agentToolComplete: "agent:tool-complete",
  terminalOutput: "terminal:output",
  terminalExit: "terminal:exit",
  projectFileChanged: "project:file-changed",
  userSpecsChanged: "user:specs-changed",
  taskUpdated: "task:updated",
  ragIndexProgress: "rag:index-progress",
  mcpEvent: "mcp-event",
} as const;

export async function listenEvent<T>(
  event: (typeof TAURI_EVENTS)[keyof typeof TAURI_EVENTS],
  handler: (payload: T) => void,
): Promise<UnlistenFn> {
  if (!isTauri()) return () => undefined;
  return listen<T>(event, ({ payload }) => handler(payload));
}

File: src-tauri/src/agent/events.rs
use serde::Serialize;
use tauri::{AppHandle, Emitter, Runtime};

pub const AGENT_MESSAGE: &str = "agent:message";
pub const AGENT_STATUS: &str = "agent:status";
pub const AGENT_TOOL_START: &str = "agent:tool-start";
pub const AGENT_TOOL_COMPLETE: &str = "agent:tool-complete";
pub const TERMINAL_OUTPUT: &str = "terminal:output";
pub const TERMINAL_EXIT: &str = "terminal:exit";
pub const PROJECT_FILE_CHANGED: &str = "project:file-changed";
pub const USER_SPECS_CHANGED: &str = "user:specs-changed";
pub const TASK_UPDATED: &str = "task:updated";

pub fn emit<R: Runtime, T: Serialize + Clone>(
    app: &AppHandle<R>,
    event: &str,
    payload: T,
) -> Result<(), String> {
    app.emit(event, payload).map_err(|error| error.to_string())
}
